### PR TITLE
#741  Parse categories

### LINKED ISSUE #741: parse_stalbeton.py:83-84: Parse stalbeton pages. See...
The puzzle `736-bb965292` from #736 has to be resolved: 

https://github.com/fidals/stroyprombeton/blob/f7e5cfc60349563bb851b7329b2b635d2dfc9f6f/stroyprombeton/management/commands/parse_stalbeton.py#L83-L84

The puzzle was created by duker33 on 10-Jul-19. 

Estimate: 120 minutes,  role: DEV. 

If you have any technical questions, don't ask me, submit new tickets instead. The task will be \"done\" when the problem is fixed and the text of the puzzle is _removed_ from the source code. Here is more about [PDD](http://www.yegor256.com/2009/03/04/pdd.html) and [about me](http://www.yegor256.com/2017/04/05/pdd-in-action.html).

### PR DESCRIPTION
**Pull request by duker33**
Closes #741

### CODE REVIEW

File: stroyprombeton/management/commands/parse_stalbeton.py [hunk 1]
```diff
@@ -9,88 +10,119 @@
...
-    #  instead of nested category parsing, drafted with the code below.
-    roots = CatalogPage().roots()  # Ignore PyFlakesBear
-    second_level = CatalogPage().second_level()
-    third_level = chain.from_iterable((s.third_level() for s in second_level))
-    options = chain.from_iterable((t.options() for t in third_level))  # Ignore PyFlakesBear
-    # save it into our DB
+    main = Page(path='/')
+    through = ThroughElements(page=main)
+    roots = through.roots()
+    seconds = chain.from_iterable((r.second_level() for r in roots))
```
Comment (artemiy312) [inline stroyprombeton/management/commands/parse_stalbeton.py]: We can make `CategoryPage` implementers polymorphic 
```python
class CategoryPage(Page):
    def children(self) -> List['CategoryPage']:
        raise NotImplementedError
    def options(self) -> list:
        return list(chain.from_iterable(child.options() for child in self.children()))
    def series(self) -> list:
        return list(chain.from_iterable(child.series() for child in self.children()))
```
And implement them in this way:
```python
class Categories(CategoryPage):
    def children(self):
        return list(chain.from_iterable(category.children() for category in self.categories))

class RootCategoryPage(CategoryPage):
    def children(self):
        return [SecondLevelCategoryPage(...) for ... in ...]
class SecondLevelCategoryPage(...):
class ThirdLevelCategoryPage(...): 
    def children(self):
        return []
    def options(self):
        # parse
```
and use it so:
```
categories = Categories(roots)
options = categories.options()
series = Categories(categories.children()[:100]).series()
```

File: stroyprombeton/management/commands/parse_stalbeton.py [hunk 2]
```diff
@@ -9,88 +10,119 @@
...
+    thirds = chain.from_iterable((s.third_level() for s in seconds))
+    options = chain.from_iterable((t.options() for t in thirds))  # Ignore PyFlakesBear
+    # @todo #741:60m Save parsed stalbeton to a DB.
+    #  It should not some high performance DB.
```
Comment (artemiy312) [inline stroyprombeton/management/commands/parse_stalbeton.py]: Why not? We already have dockerized postgres, that we don't need to setup, so we can just create a separate database for these data.

With sqlite we have to manage two difference databases in the project

I'd prefer to stay with postgres. Up to you

Comment (duker33) [reply stroyprombeton/management/commands/parse_stalbeton.py]: @artemiy312 , we'll make this decision during the substask implementation.
I've meant that DB is not required to be high performance. I'll clarify tasks body with it

File: stroyprombeton/management/commands/parse_stalbeton.py (hunk 1 above)
Comment (duker33) [reply stroyprombeton/management/commands/parse_stalbeton.py]: @artemiy312 , semantically not every category page has it's own options we should extract. The same is for series. I proposed a more rude non-polymorphic structure to clarify some points:
- we can correctly extract options and series only from third level categories
- about categories nesting: only roots can give us seconds. And only seconds - thirds.

Now i'm going to merge the task, but i'll create separated subtask to discuss it.

Categories class is a good idea, i'll subtask it

**Pull request merged**